### PR TITLE
viz: add base path for web dashboard

### LINKED ISSUE #11851: Viz Dashboard / Web serve from subpath
### What problem are you trying to solve?

When exposing the Viz dashboard via a reverse-proxy or ingress rule, it can be necessary to serve from a subpath instead of `/` directly, for instance:
`https://my-great-meshed-system/linkerd-viz`

### How should the problem be solved?

Some argument to the _web_ component like "--serve-from-subpath /linkerd-viz" to be respected in all http routes as well as rendered hyperlinks.

### Any alternatives you've considered?

Doing url-rewrite on ingress rule, but that doesn't help for self-links returned from the web component.

### How would users interact with this feature?

via the new command line argument for setting the sub-path to serve from.

### Would you like to work on this feature?

maybe

### PR DESCRIPTION
**Pull request by WatcherWhale**
Fixes #11851
Signed-off-by: Mathias Maes <mathias.maes@aloxy.io>

### CODE REVIEW

**Comment by WatcherWhale** [comment]
> grafana also needs to consider switching uri's，Is it easy to modify? I can provide a test
> 
> #12015

This should already work, as I also changed the grafana reverse proxy in the go code

File: web/app/js/index.js [hunk 1]
```diff
@@ -36,18 +36,23 @@ import { dashboardTheme } from './components/util/theme.js';
...
 }
 
+// Make sure the pathPrefix does not end with a '/'
+if (pathPrefix.slice(-1) === '/') {
+  pathPrefix = pathPrefix.slice(0, -1);
+}
+
 let defaultNamespace = 'default';
-const pathArray = window.location.pathname.split('/');
+const pathArray = window.location.pathname.substring(pathPrefix.length).split('/');
```
Comment (kflynn) [inline web/app/js/index.js:51]: It feels a little weird to me to not make sure that the path really does start with `pathPrefix` before doing this substring...

File: web/app/js/index.js [hunk 2]
```diff
@@ -36,18 +36,23 @@ import { dashboardTheme } from './components/util/theme.js';
...
 
 // if the current URL path specifies a namespace, this should become the
 // defaultNamespace
-if (pathArray[0] === '' && pathArray[1] === 'namespaces' && pathArray[2]) {
+if (pathPrefix !== '' && pathArray[0] === '' && pathArray[1] === 'namespaces' && pathArray[2]) {
```
Comment (kflynn) [inline web/app/js/index.js:55]: This doesn't seem quite right – you're making the entire check for a namespace in the path conditional on the path prefix being set?

**Review by kflynn — CHANGES_REQUESTED** [review]
Hey, my apologies for the delay looking into this! I had to figure out some stuff on my end before I could look. 😐 

I left a couple of questions on the code, but the two biggest things I noticed when testing were

1. When I don't set `-base-path` I get 404s. 🙂 
2. I didn't actually manage to get Grafana hooked in – this might be my setup, though, how did you test this?

Thanks much for getting into this! and I apologize, again, for the delay. 😐

File: viz/charts/linkerd-viz/README.md [hunk 3]
```diff
@@ -149,6 +150,7 @@ Kubernetes: `>=1.22.0-0`
 | prometheus.podAnnotations | object | `{}` | annotations for the prometheus pod |
 | prometheus.proxy | string | `nil` |  |
 | prometheus.remoteWrite | string | `nil` | Allows transparently sending samples to an endpoint. Mostly used for long term storage. |
+| prometheus.resources | object | `{"cpu":{"limit":null,"request":null},"ephemeral-storage":{"limit":"","request":""},"memory":{"limit":null,"request":null}}` | Prometheus data volume size.  size: |
```
Comment (kflynn) [inline viz/charts/linkerd-viz/README.md:153]: Ah yes, I feel like this should be a separate PR? or maybe this is just included by mistake.

**Comment by adleong** [comment]
Closing as inactive.  Please feel free to reopen if you plan to continue work on this.

**Pull request closed without merging**